lib/Value/Matrix.pm

```diff
@@ -749,27 +952,171 @@ sub column {
-#
+=head3 C<element>
+
+Extract an element from the given row/col.
+
+Usage:
+
+    my $A    = Matrix([ [ 1, 2, 3, 4 ], [ 5, 6, 7, 8 ], [ 9, 10, 11, 12 ] ]);
+    $A->element(2,3); # returns 7
+
+    my $B = Matrix([ [ [ 1, 2 ], [ 3, 4 ] ], [ [ 5, 6 ], [ 7, 8 ] ] ]);
+    $B->element(1,2,1); # returns 3;
+
+    my $row = Matrix([4,3,2,1]);
+    $row->element(2); # returns 3;
+=cut
+
 sub element {
 	my $self = (ref($_[0]) ? $_[0] : shift);
 	my $M    = $self->promote(shift);
 	return $M->extract(@_);
 }
 
-# @@@ assign @@@
+=head3 C<setElement>
+
+Assign an element in the matrix to a value.
+
+Inputs: indices as an arrayref and the value in a form that can be parsed.
+
+Note: this mutates the matrix itself.
+
+Usage:
+
+    my $A = Matrix([ [ 1, 2, 3, 4 ], [ 5, 6, 7, 8 ], [ 9, 10, 11, 12 ] ]);
+    $A->setElement([2,3],-5);
+
+
+=cut
+
+sub setElement {
+	my ($self, $ind, $value) = @_;
+
+	# Drill down into the matrix
+	my $el = $self->{data}[ $ind->[0] - 1 ];
+	for my $i (1 .. scalar(@$ind) - 1) { $el = $el->{data}[ $ind->[$i] - 1 ]; }
+
+	# update the value of $el
+	$el = Value::makeValue($value);
+}
+
+=head3 C<subMatrix>
+
+
+Return a submatrix of the matrix.  If the rows and columns are array refs, the given rows and
+columns of the matrix are returns as a Matrix object.
+
+If the input are integers, then the submatrix with that row and column removed.
+
+Usage:
+
+    $A = Matrix([[1,2,3,4],[5,6,7,8],[9,10,11,12]]);
+
+    $A->subMatrix([2..3],[2..4]);  # returns a Matrix([[6,7,8],[10,11,12]])
+
+    $A->subMatrix(2,3);  # returns Matrix([ [ 1, 2, 4 ], [ 9, 10, 12 ] ]);
+
+    $A->subMatrix([3,1,2],[1,4,2]);  # returns Matrix([9,12,10],[1,4,2],[5,8,6]);
+=cut
+
+sub subMatrix {
+	my ($self, $r, $c) = @_;
+	my $context = $self->context;
+	my ($rows, $cols);
+	my ($nrow, $ncol) = $self->dimensions;
+
+	# check if the inputs are integers.
+	if (ref $r eq '' && ref $c eq '') {
+		Value::Error("The input $r is not a valid row.")    unless $r >= 1 && $r <= $nrow && int($r) == $r;
+		Value::Error("The input $c is not a valid column.") unless $c >= 1 && $c <= $ncol && int($c) == $c;
+		$rows = [ grep { $_ != $r } (1 .. $nrow) ];
+		$cols = [ grep { $_ != $c } (1 .. $ncol) ];
+	} elsif (ref $r eq 'ARRAY' && ref $c eq 'ARRAY') {
+		$rows = $r;
+		$cols = $c;
+		for my $i (@$rows) {
+			Value::Error("The input $i is not a valid row.") unless int($i) == $i && $i >= 1 && $i <= $nrow;
+		}
+		for my $i (@$cols) {
+			Value::Error("The input $i is not a valid column.") unless int($i) == $i && $i >= 1 && $i <= $ncol;
+		}
+	} else {
+		Value::Error('The inputs must both be integers or array refs.');
+	}
+	my @M = ();
+	for my $r (@$rows) {
+		push(@M, $self->make($context, map { $self->element($r, $_) } @$cols));
+	}
+	return $self->make($context, @M);
+}
+
+=head3 C<removeRow>
+
+Remove a row from a matrix.
+
+Usage:
+
+    $A = Matrix([ [ 1, 2, 3, 4 ], [ 5, 6, 7, 8 ], [ 9, 10, 11, 12 ], [13, 14, 15, 16] ]);
+    $A->removeRow(3);
```

Comment:
This method isn't working.  I used your example above, and it returned the matrix `[[1],[2],[3]]`.

Reply:
Also, since `$A` is not mutated, your POD example should assign the return result to something.